crates/node-builder/src/exex.rs

```diff
@@ -0,0 +1,96 @@
+#![allow(dead_code)]
+
+use crate::FullNodeTypes;
+use futures::{future::BoxFuture, Future, FutureExt, Stream};
+use reth_node_core::{
+    dirs::{ChainPath, DataDirPath},
+    node_config::NodeConfig,
+};
+use reth_primitives::{BlockNumber, Head};
+use reth_tasks::TaskExecutor;
+use std::{borrow::Cow, pin::Pin};
+
+/// An ExEx (Execution Extension) that processes new blocks and emits events on a stream.
+pub trait ExEx: Stream<Item = ExExEvent> + Send + 'static {
+    /// Returns the name of the ExEx. It will appear in logs and metrics.
+    fn name(&self) -> Cow<'static, str>;
+}
+
+/// Events emitted by an ExEx.
+#[derive(Debug)]
+pub enum ExExEvent {
+    /// Highest block processed by the ExEx.
+    ///
+    /// ExEx must guarantee that it will not require all earlier blocks in the future, meaning
+    /// that Reth is allowed to prune them.
+    ///
+    /// On reorgs, it's possible for the height to go down.
+    FinishedHeight(BlockNumber),
+}
+
+/// Captures the context that an ExEx has access to.
+#[derive(Debug)]
+pub struct ExExContext<Node: FullNodeTypes> {
+    /// The current head of the blockchain at launch.
+    head: Head,
+    /// The configured provider to interact with the blockchain.
+    provider: Node::Provider,
+    /// The executor of the node.
+    executor: TaskExecutor,
+    /// The data dir of the node.
+    data_dir: ChainPath<DataDirPath>,
+    /// The config of the node
+    config: NodeConfig,
+    /// loaded config
+    reth_config: reth_config::Config,
+    // TODO(alexey): add pool, payload builder, anything else?
```

Comment:
it's in the provider, but unsure if we can reuse it, because we'll not just use blockchain tree's state notifications, but also the ones coming from the Execution stage

Reply:
yeah, we will want our own notif stream, but i assume we are adding that to this struct at some point? since its not present in the trait

Reply:
yep, it will be in the struct

Reply:
this is available throught the `Provider`, but if we need an extra event for pipeline sync this is where we'd unify them

Reply:
p2p network? such that i can direct broadcast a message to ethereum p2p as an exec? because if I'm in a discv5 world this allows ExEx's to communicate over p2p topics